### ISSUE TITLE
compiling with default file values

### ISSUE THREAD

**Issue by dpark01** [dnanexus/dxWDL#69]
Hi again,

I'm trying out the `-defaults` option for dxWDL `compile` and I am attempting to set up the json as described in the dxWDL readme:

```
{
  "assemble_denovo_with_deplete.deplete_taxa.bmtaggerDbs": [
    "dx://file-BYF8y0Q06PJ7G1fPvkB9q3fK",
    "dx://file-BYF6g8Q0zjF77x79bGYgJ1Zb",
    "dx://file-BYF6j80068477x79bGYgJ1gX"
  ],
  "assemble_denovo_with_deplete.deplete_taxa.blastDbs": [
    "dx://file-BzBpkK80x0z8GBQPfZ4kgPKp",
    "dx://file-BYFv7bQ0QyBFXQBkX1Jqz88g"
  ],

  "assemble_denovo_with_deplete.assemble.trim_clip_db":
    "dx://file-BXF0vYQ0QyBF509G9J12g927"
}
```

These file IDs refer to objects in the target project that I am compiling to. If I try to compile with the `-defaults` option, I get this error from dxWDL:

```
dxWDL.AppInternalException: java.lang.RuntimeException: File transfer not allowed {
  "$dnanexus_link": "file-BYF8y0Q06PJ7G1fPvkB9q3fK"
} direction=Zero
```

I tried prepending the `file-` with a `project-` just for good measure and got this error instead:
```
lookupObject: project-F856jv809y3VkzyFGkKqX367:file-BYF8y0Q06PJ7G1fPvkB9q3fK
java.lang.Exception: Object lookup requires project context
```

The three parameters I'm trying to set are an `Array[File]`, `Array[File]`, and `File?`, respectively.

Am I doing this wrong?

Also, is a Github issue the right place for general Q&A like this?

**Comment by orodeh** [comment]
Since we don't have a forum, or a google group, this is currently the best place to describe problems you run into with dxWDL. 

You are doing this exactly right. The problem is the file type, which is handled incorrectly in this case. Converting between DNAx files WDL files is tricky, and done differently depending on context. Here, we don't have a runtime environment, so we can't upload or download files, hence the `File transfer not allowed` error.

I will work on a fix, and keep you posted. Thanks for the accurate and easily reproducible bug report.

**Comment by dpark01** [comment]
Thanks... I'm a little confused though. I pre-staged these files to the project manually, and those should be the proper DNAx file identifiers, so I'm not sure why dxWDL thinks it is necessary to transfer anything.

**Comment by orodeh** [comment]
The core issue is that WDL works with the `wdl:file` type, and DNAx works with `dx:file` type. WDL files are paths in a virtual file-system (`/A/B/C`). DNAnexus has a cloud based object-store, where files are identified by ids like `file-xxxxyyyyzzzz`. These are two very different systems.

When you set file defaults, we are currently converting the file-id to a path, to get a valid WDL variable. This can only be done at **runtime**, because you need to download the file to a local path in a cloud instance. At compile time, we don't have a runtime system, we don't know which cloud instance will be used, and we can't determine the path.

I think it is possible to avoid the file-id to file-path conversion, but I need to think about it carefully. 

I hope this sheds some light on the issue.

**Comment by dpark01** [comment]
Hi sorry, I'm still not quite following. Perhaps you're saying that the WDL `File` type is not the opaque string I thought it was, but is forced to follow some sort of convention, but I still don't see the barrier.

Although many of my WDLs may include a default value for a `File` parameter specified as a GCS object (e.g. `gs://bucket/object`), which is obviously biased towards the GCE execution environment (but not necessarily), I can execute these same WDLs on a local execution environment by providing local file paths as well for the input.json. In other words, my understanding of a WDL `File` is that you can provide inputs or defaults using whatever identifiers or URI schemas are relevant to your execution environment. Based on the dxWDL documentation, I just assumed I could pass inputs (or specify default values) using DNAnexus file IDs like `file-xxxx` or, as the dxWDL documentation implies, `dx:file-xxxx`.

Is `dx:file-xxxx` somehow not a valid WDL `File` variable? Does wdl4s somehow unfairly impose some required format on the File string that it probably shouldn't be?

Converting the file-id to a path is not really any different than what Google Pipelines does when it localizes its files based on `gs://xxx` strings, and of course, it's done at runtime, but that doesn't prevent anyone from setting a default value for the WDL. I'm not seeing why dxWDL can't do the same?

Basically, my expectation is that if I compile a WDL workflow with dxWDL and provide a `file-XXX` or `dx:file-XXX` in a `-defaults` json file during compile time, it would result in a compiled workflow that looked something like this:

```
 "inputSpec": [
    {
      "name": "parameter_name",
      "class": "array:file",
      "default": [
        {"$dnanexus_link": "file-XXX"},
```

Example of desired output here: https://github.com/dnanexus/viral-ngs/blob/dnanexus/assembly/viral-ngs-human-depletion/dxapp.json#L28

Again, it can be left to the user of dxWDL to pre-stage the files to DNAnexus and provide the appropriate DNAnexus file IDs at compile time.

**Comment by orodeh** [comment]
This is definitely supposed to work in exactly the way you understand. 

This is a dxWDL that I need to fix, and I was trying to explain the internals.  

I am actually working on this case, and a few related cases, having to do with converting between DNAx and WDL files. 

I hope this is clearer.

**Comment by orodeh** [comment]
@dpark01, I believe release 0.52 fixes this problem. Please give it a try, and let me know if the problem persists.

**Comment by dpark01** [comment]
Okay, my ability to evaluate this might be impacted a bit by #80 but here's what I noticed.

 - dxWDL compile workflow.wdl -defaults defaults.json succeeds now
 - `dx run --help workflow-xxxx` on the compiled output looks proper. I see the input `File` and `Array[File]` inputs and they do appear to be defaulted properly (e.g. `assemble___trim_clip_db: [-iassemble___trim_clip_db=(file, default={"$dnanexus_link": "file-BXF0vYQ0QyBF509G9J12g927"})]`)
 - All of the inputs that were given default values during compile are invisible in the web UI -- I cannot change these values or even see that they ever existed when trying to launch an analysis. This is a significant problem.

Also, is there currently a mechanism to specify a default File as `project-xx:file-yy`? I tried and the compiler wasn't happy.

**Comment by orodeh** [comment]
I hear you regarding the UI, we will have to see what can be done. 

If you want to specify a file by ID, this should work: `dx://project-xxxx:file-yyyy`.

**Comment by orodeh** [comment]
@dpark01, has this been fixed in release 0.54?

**Comment by dpark01** [comment]
Sorry, I didn't have much time to dig into this at the end of last week, but I have some observations now.

Here is a [WDL workflow](https://github.com/broadinstitute/viral-ngs/blob/dp-dx-update/pipes/WDL/workflows/assemble_denovo_with_deplete.wdl) (a simple linear workflow) that I compiled with these [json defaults](https://github.com/broadinstitute/viral-ngs/blob/dp-dx-update/pipes/WDL/dx-defaults-assemble_denovo_with_deplete.json) using dxWDL 0.54.  The result can be found in `project-F856jv809y3VkzyFGkKqX367:workflow-F8g9K8009y3zj6pX7gKQxVZY`.

The compile succeeds. If I `dx run --help` the above workflow, I can see that those files have default values that appear to be properly formed dnanexus_link objects. However, in the web UI, when I try to launch the workflow, there is no visual indication that these parameters have any default values assigned. They (visually) appear to be empty/unassigned inputs.

I haven't yet tried executing them via web UI to see if it uses the default values anyway.

By contrast, here is a (@mlin created non-dxWDL) DNAnexus workflow in `project-BXx29vQ0y3FfPJv100237B88:workflow-F7PyX980y3FyXzYvG5QFZQz2` and it displays its default values for the same parameters (bmtagger_dbs, blast_dbs, contaminants) in a similar manner with `dx run --help` but very differently in the web UI launcher. In this case, it clearly states that bmtagger_dbs has "3 inputs" populated and blast_dbs has "2 inputs" populated, and the contaminants input shows the actual filename of the defaulted input fasta file.

While I'm not yet sure if there is any functional difference, do you have any guess as to why there's a UI difference between the dxWDL output and the natively-created workflow?

**Comment by dpark01** [comment]
FYI, I did try executing the compiled workflow from the web UI and I didn't specify any inputs to the defaulted File and Array[File] parameters and that did work properly (well, the job failed for reasons that had nothing to do with this, but it did set the inputs to the defaults I asked for)

**Comment by orodeh** [comment]
@dpark01, I managed to look at the problem, and view it on my end. I am talking with the UI folks to try to figure it out.

**Comment by dpark01** [comment]
This looks mostly working to me in 0.56, but, as described in a separate email, I'm having trouble figuring out how to name input parameters that are within scatter blocks. It may be worth adding a bit of documentation in the README about how implicit inputs get named for the inputs json and defaults json.

**Comment by dpark01** [comment]
This works in 0.57!
